gossip/src/received_cache.rs

```diff
@@ -0,0 +1,185 @@
+        }
+        .into_iter()
+        .flatten()
+    }
+
+    // Only for tests and simulations.
+    pub(crate) fn mock_clone(&self) -> Self {
+        let mut cache = LruCache::new(self.0.cap());
+        for (&origin, entry) in self.0.iter().rev() {
+            cache.put(origin, entry.clone());
+        }
+        Self(cache)
+    }
+}
+
+impl ReceivedCacheEntry {
+    // Limit how big the cache can get if it is spammed
+    // with old messages with random pubkeys.
+    const CAPACITY: usize = 50;
+    // Threshold for the number of duplicates before which a message
+    // is counted as timely towards node's score.
+    const NUM_DUPS_THRESHOLD: usize = 2;
+
+    fn record(&mut self, node: Pubkey, num_dups: usize) {
+        if num_dups == 0 {
+            self.num_upserts = self.num_upserts.saturating_add(1);
+        }
+        // If the message has been timely enough increment node's score.
+        if num_dups < Self::NUM_DUPS_THRESHOLD {
+            let score = self.nodes.entry(node).or_default();
+            *score = score.saturating_add(1);
+        } else if self.nodes.len() < Self::CAPACITY {
+            // Ensure that node is inserted into the cache for later pruning.
+            let _ = self.nodes.entry(node).or_default();
+        }
+    }
+
+    fn prune(
+        self,
+        pubkey: &Pubkey, // This node.
+        origin: &Pubkey, // CRDS value owner.
+        stake_threshold: f64,
+        min_ingress_nodes: usize,
+        stakes: &HashMap<Pubkey, u64>,
+    ) -> impl Iterator<Item = Pubkey> {
+        debug_assert!(self.num_upserts >= ReceivedCache::MIN_NUM_UPSERTS);
+        // Enforce a minimum aggregate ingress stake; see:
+        // https://github.com/solana-labs/solana/issues/3214
+        let min_ingress_stake = {
+            let stake = stakes.get(pubkey).min(stakes.get(origin));
+            (stake.copied().unwrap_or_default() as f64 * stake_threshold) as u64
+        };
+        self.nodes
+            .into_iter()
+            .map(|(node, score)| {
+                let stake = stakes.get(&node).copied().unwrap_or_default();
+                (node, score, stake)
+            })
+            .sorted_unstable_by_key(|&(_, score, stake)| Reverse((score, stake)))
+            .scan(0u64, |acc, (node, _score, stake)| {
+                let old = *acc;
+                *acc = acc.saturating_add(stake);
+                Some((node, old))
+            })
+            .skip(min_ingress_nodes)
+            .skip_while(move |&(_, stake)| stake < min_ingress_stake)
+            .map(|(node, _stake)| node)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use {super::*, std::collections::HashSet};
+
+    #[test]
+    fn test_received_cache() {
+        let mut cache = ReceivedCache::new(/*capacity:*/ 100);
+        let pubkey = Pubkey::new_unique();
+        let origin = Pubkey::new_unique();
+        let nodes: Vec<_> = std::iter::repeat_with(Pubkey::new_unique).take(5).collect();
+        let records = vec![
```

Comment:
could flip these declarations and use `records.len()` instead of managing the magic number

Reply:
done.